mypyc/emitmodule.py

```diff
@@ -216,62 +304,151 @@ def generate_c_for_modules(self) -> List[Tuple[str, str]]:
 
-        return file_contents + [('__native.c', ''.join(emitter.fragments)),
-                                ('__native_internal.h', ''.join(declarations.fragments)),
-                                ('__native.h', ''.join(ext_declarations.fragments)),
+        return file_contents + [('__native{}.c'.format(self.group_suffix),
+                                 ''.join(emitter.fragments)),
+                                ('__native_internal{}.h'.format(self.group_suffix),
+                                 ''.join(declarations.fragments)),
+                                ('__native{}.h'.format(self.group_suffix),
+                                 ''.join(ext_declarations.fragments)),
                                 ]
 
+    def generate_export_table(self, decl_emitter: Emitter, code_emitter: Emitter) -> None:
+        """Generate the declaration and definition of the group's export struct.
+
+        To avoid needing to deal with deeply platform specific issues
+        involving dynamic library linking (and some possibly
+        insurmountable issues involving cyclic dependencies), compiled
+        code accesses functions and data in other compilation groups
+        via an explicit "export struct".
+
+        Each group declares a struct type that contains a pointer to
+        every function and static variable it exports. It then
+        populates this struct and stores a pointer to it in a capsule
+        stored as an attribute named 'exports' on the group's shared
+        library's python module.
+
+        On load, a group's init function will import all of its
+        dependencies' exports tables using the capsule mechanism and
+        copy the contents into a local copy of the table (to eliminate
+        the need for a pointer indirection when accessing it).
+
+        Then, all calls to functions in another group and accesses to statics
+        from another group are done indirectly via the export table.
+        """
+
+        decls = decl_emitter.context.declarations
+
+        decl_emitter.emit_lines(
+            '',
+            'struct export_table{} {{'.format(self.group_suffix),
+        )
+        for name, decl in decls.items():
+            if decl.needs_export:
+                decl_emitter.emit_line(pointerize('\n'.join(decl.decl), name))
+
+        decl_emitter.emit_line('};')
+
+        code_emitter.emit_lines(
+            '',
+            'static struct export_table{} exports = {{'.format(self.group_suffix),
+        )
+        for name, decl in decls.items():
+            if decl.needs_export:
+                code_emitter.emit_line('&{},'.format(name))
+
+        code_emitter.emit_line('};')
+
     def generate_shared_lib_init(self, emitter: Emitter) -> None:
         """Generate the init function for a shared library.
 
-        A shared library contains all of the actual code for a set of modules.
+        A shared library contains all of the actual code for a
+        compilation group.
```

Comment:
A more concrete explanation would be helpful. For each module in a group, we generate a shim with a name derived from the module name (?) and a shared library with a name that is not derived in a straightforward way from the module name, right? But if the group only has one module, do we need a separate shared lib?

Reply:
We need the separate shared lib for an annoying and perhaps somewhat subtle reason: the reading of export tables from dependencies is done as a separate phase from actually executing module top levels, because I wanted to make sure that the export tables were always initialized and couldn't get mixed up trouble due to import cycles.

But:
 1. I think I have figured out a way to keep that property while allowing just one extension module per python module, though it is somewhat unpleasant
 2. Maybe it isn't actually important anyway

I'll document what's going on here a bit more and consider one of those alternatives in a follow-up maybe